calyx/src/analysis/live_range_analysis.rs

```diff
@@ -9,44 +9,55 @@ use std::{
     ops::{BitOr, Sub},
 };
 
+type LiveCellRepresentation = HashSet<(ir::CellType, ir::Id)>;
+const NODE_ID: &str = "NODE_ID";
+
 /// The data structure used to represent sets of ids. This is used to represent
 /// the `live`, `gen`, and `kill` sets.
 #[derive(Default, Clone)]
 pub struct Prop {
-    set: HashSet<ir::Id>,
+    map: HashMap<ir::CellType, HashSet<ir::Id>>,
 }
 
 /// Conversion to Prop from things that can be converted to HashSet<ir::Id>.
-impl<T: Into<HashSet<ir::Id>>> From<T> for Prop {
+impl<T: Into<HashMap<ir::CellType, HashSet<ir::Id>>>> From<T> for Prop {
     fn from(t: T) -> Self {
-        Prop { set: t.into() }
+        Prop { map: t.into() }
     }
 }
 
 /// Implement convenience math operators for Prop
 impl BitOr<&Prop> for &Prop {
     type Output = Prop;
     fn bitor(self, rhs: &Prop) -> Self::Output {
-        Prop {
-            set: &self.set | &rhs.set,
+        let mut map: HashMap<_, HashSet<_>> = self.map.clone();
```

Comment:
This clone seems expensive. Maybe there is a way to iterate over the `self` map while using `remove` calls on the other HashMap to take ownership of the values associated with a key.